extensions/pkg/controller/worker/helper/helper.go

```diff
@@ -77,3 +82,29 @@ func GetMachineSetWithMachineClass(machineDeploymentName, machineClassName strin
 	}
 	return nil
 }
+
+// ReportFailedMachines reports the names of failed machines in the given `status` per error description.
+func ReportFailedMachines(status machinev1alpha1.MachineDeploymentStatus) error {
+	machines := status.FailedMachines
+	if len(machines) == 0 {
+		return nil
+	}
+
+	descriptionPerFailedMachines := make(map[string]string)
+	for _, machine := range machines {
+		if _, ok := descriptionPerFailedMachines[machine.LastOperation.Description]; !ok {
+			descriptionPerFailedMachines[machine.LastOperation.Description] = fmt.Sprintf("%q", machine.Name)
+			continue
+		}
+		descriptionPerFailedMachines[machine.LastOperation.Description] += fmt.Sprintf(", %q", machine.Name)
+	}
+
+	allErrs := &multierror.Error{
+		ErrorFormat: utilerrors.NewErrorFormatFuncWithPrefix("machine(s) failed"),
+	}
+	for description, names := range descriptionPerFailedMachines {
+		allErrs = multierror.Append(allErrs, fmt.Errorf("%s: %s", names, description))
+	}
```

Comment:
What about replacing `map[string]string` with `map[string][]string` and `append`ing the machine names, so that you can later `strings.Join` them with `,` as delimiter? This way you also wouldn't need the check in line 95 whether the description is already known.

Reply:
Sure, no strong opinion.